src/java/org/apache/cassandra/utils/memory/MemtableCleanerThread.java

```diff
@@ -27,61 +33,118 @@
-                cleaner.run();
-            }
-            else
+            if (!pool.needsCleaning())
             {
                 final WaitQueue.Signal signal = wait.register();
-                if (!needsCleaning())
+                if (!pool.needsCleaning())
                     signal.await();
                 else
                     signal.cancel();
             }
+            else
+            {
+                // Increment the number of pending tasks and determine the minimum ownership ratio to pass to the clener
+                // This is zero (no min) if we have less than maxPendingTasks, otherwise it is the cleanThreshold
+                // The reason is that we don't want to queue more than numPendingTasks, but at the same time we don't
+                // want to end up with memtables that are too large, so if there is a memtable larger than cleanThreshold,
+                // we want to flush that
+                int numPendingTasks = this.numPendingTasks.incrementAndGet();
+                double minOwnershipRatio = numPendingTasks > maxPendingTasks ? cleanThreshold : 0;
+
+                if (logger.isTraceEnabled())
+                    logger.trace("Invoking cleaner with {} tasks pending", numPendingTasks);
+
+                cleaner.clean(minOwnershipRatio).handle(this::apply);
+            }
+        }
+
+        private Boolean apply(Boolean res, Throwable err)
+        {
+            final int tasks = this.numPendingTasks.decrementAndGet();
+
+            // if the cleaning job was scheduled (res == true) or had an error, trigger again after decrementing the tasks
+            // Even if the pool also triggers when task completes (memtable reclaimed), this new trigger
+            // may be able to reclaim more memory by passing a lower min ownership ratio because the pending tasks
+            // have been decremented
+            if ((res || err != null) && pool.needsCleaning())
+                wait.signal();
+
+            if (err != null)
+                logger.error("Memtable cleaning tasks failed with an exception and {} pending tasks ", tasks, err);
+            else if (logger.isTraceEnabled())
+                logger.trace("Memtable cleaning task completed ({}), currently pending: {}", res, tasks);
+
+            return res;
         }
     }
 
     private final Runnable trigger;
+    private final Clean<P> clean;
+
     private MemtableCleanerThread(Clean<P> clean)
     {
         super(clean.pool.getClass().getSimpleName() + "Cleaner", clean);
         this.trigger = clean.wait::signal;
+        this.clean = clean;
     }
 
-    MemtableCleanerThread(P pool, Runnable cleaner)
+    MemtableCleanerThread(P pool, MemtableCleaner cleaner, int maxPendingTasks, double cleanThreshold)
     {
-        this(new Clean<>(pool, cleaner));
+        this(new Clean<>(pool, cleaner, maxPendingTasks, cleanThreshold));
     }
 
     // should ONLY be called when we really think it already needs cleaning
```

Comment:
This could be JavaDoc instead of a line comment:
```suggestion
    /** Should ONLY be called when we really think it already needs cleaning */
```